### PR TITLE
#1030 [RTC] Implement executeAsyncScript

### PR DESCRIPTION
**Pull request by tyge68**
## Proposed changes
Adding executeAsyncJavascript(...) to the static Selenide object.
Adding new integration tests in the SelenideMethodsTest
Adding implementation in the DriverStub for any other tests to eventually use it as well

## Checklist
- [ ] Checkstyle and unit tests pass locally with my changes by running `gradle check chrome_headless firefox_headless` command
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: statics/src/main/java/com/codeborne/selenide/Selenide.java [hunk 1]
```diff
@@ -582,10 +582,20 @@ public static ElementsCollection getElements(By criteria) {
     return getSelenideDriver().findAll(criteria);
   }
 
+  /**
+   * @see @see <a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html#executeScript-java.lang.String-java.lang.Object...-">executeScript</a>
```
Comment (rosolko) [inline statics/src/main/java/com/codeborne/selenide/Selenide.java]:
```suggestion
   * @see <a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html#executeScript-java.lang.String-java.lang.Object...-">executeScript</a>
```

Looks like checkstyle isn't pass such long line. We have a limit equals to `136` characters.

File: statics/src/main/java/com/codeborne/selenide/Selenide.java [hunk 2]
```diff
@@ -582,10 +582,20 @@ public static ElementsCollection getElements(By criteria) {
...
+   */
   public static <T> T executeJavaScript(String jsCode, Object... arguments) {
     return getSelenideDriver().executeJavaScript(jsCode, arguments);
   }
 
+  /**
+   * @see @see <a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html#executeAsyncScript-java.lang.String-java.lang.Object...-">executeAsyncScript</a>
```
Comment (rosolko) [inline statics/src/main/java/com/codeborne/selenide/Selenide.java]:
```suggestion
   * @see <a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html#executeAsyncScript-java.lang.String-java.lang.Object...-">executeAsyncScript</a>
```

Same here.

**Comment by rosolko** [comment]
@tyge68 Please execute locally gradle  `check` task

**Comment by tyge68** [comment]
@rosolko thanks for the info, I changed to use a bit.ly short url as it is already used in this class for getWebDriverLogs(...) which avoid then to have line breaks. I also re-run "gradle check" and it passed.

Let me know if that works for you now.

**Comment by rosolko** [comment]
@tyge68 bitly links isn't a good idea, we think about them previously. 
We have several options how to avoid them:
1. Replace link with text from doc
2. Use full link and ignore it on check
3. Try to split link to multiple row if possible.
4. Extend line length in a rules
@asolntsev ^

File: statics/src/main/java/com/codeborne/selenide/Selenide.java [hunk 3]
```diff
@@ -582,10 +582,22 @@ public static ElementsCollection getElements(By criteria) {
     return getSelenideDriver().findAll(criteria);
   }
 
+  /**
+   * @see <a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html
```
Comment (asolntsev) [inline statics/src/main/java/com/codeborne/selenide/Selenide.java]: A proper way to refer other Java classes in javadoc is mentioning class name + method name with argument types:

```java
/**
 * @see org.openqa.selenium.By#className(java.lang.String)
 */
```

This way we don't need to bother about long or changing URLs.

File: statics/src/test/java/integration/SelenideMethodsTest.java [hunk 4]
```diff
@@ -564,6 +565,20 @@ void canExecuteCustomCommand() {
...
+  @Test
+  void canExecuteJavascript() {
+    long value = (Long) Selenide.executeJavaScript("return 10;");
+    assertThat(value).isEqualTo(10);
+  }
+
+  @Test
+  void canExecuteAsyncJavascript() {
+    long value = (Long) Selenide.executeAsyncJavaScript(
+      "var callback = arguments[0]; setTimeout(function() { callback(10); }, 50);"
```
Comment (asolntsev) [inline statics/src/test/java/integration/SelenideMethodsTest.java]: The proper way is `arguments[arguments.length - 1]` instead of `arguments[0]`.

Comment (tyge68) [reply statics/src/test/java/integration/SelenideMethodsTest.java]: done in my next commit

File: statics/src/main/java/com/codeborne/selenide/Selenide.java (hunk 3 above)
Comment (tyge68) [reply statics/src/main/java/com/codeborne/selenide/Selenide.java]: true,  but then you would need to find out the page in google, the idea is to provide a link still to help the user.

I suggest to add an ignore pattern in the checkstyle related rule in my next commit, that allow longer line for " * @see " which is then more url friendly. That also permit to include a longer text inside the link itself.

**Comment by tyge68** [comment]
> For example, we could introduce a "reserved word" like CALLBACK or RETURN or EXIT

I suppose you would like that such reserved keywords are replaced before doing the call to JavascriptExecutor method ?  That would be indeed a possibility. But for sure a bit outside the initial scope of this RTC. But if that is only this , why not adding this along.

File: statics/src/main/java/com/codeborne/selenide/Selenide.java (hunk 3 above)
Comment (asolntsev) [reply statics/src/main/java/com/codeborne/selenide/Selenide.java]: @tyge68 No, you don't need to google. IDEA can automatically download sources jar from maven central repo. You can easily browse sources/javadoc of any open-sources projects right in IDEA (not sure about Eclipse/NetBeans).

**Comment by asolntsev** [comment]
@tyge68 
1. Yes, I meant that reserved keywords are replaced before doing the call to JavascriptExecutor.
2. You are right, my suggestion with CALLBACK is outside of the initial scope. We can discuss it later because it doesn't influence the API. 

Rather I was thinking about creating a better API **before** we release this feature. Thought, it seems that none of us has a good idea how to improve API. So we can release it as is.

File: statics/src/main/java/com/codeborne/selenide/Selenide.java (hunk 3 above)
Comment (rosolko) [reply statics/src/main/java/com/codeborne/selenide/Selenide.java]: +

**Pull request merged**